test/src/worker.rs

```diff
@@ -0,0 +1,235 @@
+                            node_dirs,
+                        })
+                        .unwrap();
+                    return;
+                }
+                outbox
+                    .send(Notify::Done {
+                        spec_name,
+                        seconds: now.elapsed().as_secs(),
+                    })
+                    .unwrap();
+            }
+        })
+    }
+}
+
+fn run_spec(spec: &dyn crate::specs::Spec, net: &mut Net) -> ::std::thread::Result<()> {
+    panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        spec.init_config(net);
+    }))?;
+
+    panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        spec.before_run(net);
+    }))?;
+
+    spec.start_node(net);
+
+    panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        spec.run(net);
+    }))
+}
+
+/// A group of workers
+pub struct Workers {
+    workers: Vec<(Sender<Command>, Worker)>,
+    join_handles: Option<Vec<JoinHandle<()>>>,
+    is_shutdown: bool,
+}
+
+impl Workers {
+    /// Create n workers
+    pub fn new(
+        count: usize,
+        tasks: Arc<Mutex<Tasks>>,
+        outbox: Sender<Notify>,
+        start_port: u16,
+        binary: String,
+        vendor: PathBuf,
+    ) -> Self {
+        let workers: Vec<_> = (0..count)
+            .map({
+                let tasks = Arc::clone(&tasks);
+                move |i| {
+                    let (command_tx, command_rx) = unbounded();
+                    // assign 20 port slot to each spec
+                    let start_port = start_port + (i * 20) as u16;
+                    let worker = Worker::new(
+                        Arc::clone(&tasks),
+                        command_rx,
+                        outbox.clone(),
+                        start_port,
+                        binary.to_string(),
+                        vendor.clone(),
+                    );
+                    (command_tx, worker)
+                }
+            })
+            .collect();
+        Workers {
+            workers,
+            join_handles: None,
+            is_shutdown: false,
+        }
+    }
+
+    /// start all workers
+    pub fn start(&mut self) {
+        let mut join_handles = Vec::new();
+        for w in &mut self.workers {
+            let h = w.1.start();
+            join_handles.push(h);
+        }
+        self.join_handles.replace(join_handles);
+    }
+
+    /// shutdown all workers, must call join_all after this.
+    pub fn shutdown(&mut self) {
+        if self.is_shutdown {
+            return;
+        }
+        for w in &self.workers {
+            if let Err(err) = w.0.send(Command::Shutdown) {
+                info!("shutdown worker failed, error: {}", err);
+            }
+        }
+        self.is_shutdown = true;
+    }
+
+    /// wait all workers to shutdown
+    pub fn join_all(&mut self) {
+        if self.join_handles.is_none() {
+            return;
+        }
+        // make sure shutdown all workers
+        self.shutdown();
```

Comment:
why call `shutdown` before `join`?

Reply:
https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.join

Need to use join to free threads resource.